File: code/5._Huffman_Coding/Java/programming.java/Huffman_Coding.java
import java.util.PriorityQueue;
import java.util.HashMap;
import java.util.Map;

class HuffmanNode implements Comparable<HuffmanNode> {
    char data;
    int frequency;
    HuffmanNode left, right;

    public int compareTo(HuffmanNode node) {
        return frequency - node.frequency;
    }
}

public class HuffmanCoding {
    public static HuffmanNode buildHuffmanTree(Map<Character, Integer> freqMap) {
        PriorityQueue<HuffmanNode> heap = new PriorityQueue<>();

        for (Map.Entry<Character, Integer> entry : freqMap.entrySet()) {
            HuffmanNode node = new HuffmanNode();
            node.data = entry.getKey();
            node.frequency = entry.getValue();
            heap.add(node);
        }

        while (heap.size() > 1) {
            HuffmanNode left = heap.poll();
            HuffmanNode right = heap.poll();

            HuffmanNode merged = new HuffmanNode();
            merged.frequency = left.frequency + right.frequency;
            merged.left = left;
            merged.right = right;

            heap.add(merged);
        }

        return heap.poll();
    }

    public static void generateHuffmanCodes(HuffmanNode root, String code, Map<Character, String> huffmanCodes) {
        if (root == null)
            return;
        if (root.left == null && root.right == null) {
            huffmanCodes.put(root.data, code);
        }
        generateHuffmanCodes(root.left, code + "0", huffmanCodes);
        generateHuffmanCodes(root.right, code + "1", huffmanCodes);
    }

    public static String huffmanCoding(String text) {
        Map<Character, Integer> freqMap = new HashMap<>();
        for (char c : text.toCharArray()) {
            freqMap.put(c, freqMap.getOrDefault(c, 0) + 1);
        }

        HuffmanNode root = buildHuffmanTree(freqMap);
        Map<Character, String> huffmanCodes = new HashMap<>();
        generateHuffmanCodes(root, "", huffmanCodes);

        StringBuilder encodedText = new StringBuilder();
        for (char c : text.toCharArray()) {
            encodedText.append(huffmanCodes.get(c));
        }

        return encodedText.toString();
    }

    public static void main(String[] args) {
        String text = "this is an example for huffman encoding";
        String encodedText = huffmanCoding(text);
        System.out.println("Huffman Codes:");
        for (Map.Entry<Character, String> entry : huffmanCodes.entrySet()) {
            System.out.println(entry.getKey() + ": " + entry.getValue());
        }
        System.out.println("Encoded Text: " + encodedText);
    }
}

File: code/5._Huffman_Coding/Python/programming.py/Huffman_Coding.py
import heapq
from collections import defaultdict


# Node class to build the Huffman tree
class Node:
    def __init__(self, char, freq):
        self.char = char
        self.freq = freq
        self.left = None
        self.right = None

    def __lt__(self, other):
        return self.freq < other.freq


# Function to build the Huffman tree
def build_huffman_tree(freq_map):
    heap = [Node(char, freq) for char, freq in freq_map.items()]
    heapq.heapify(heap)

    while len(heap) > 1:
        left = heapq.heappop(heap)
        right = heapq.heappop(heap)
        merged = Node(None, left.freq + right.freq)
        merged.left = left
        merged.right = right
        heapq.heappush(heap, merged)

    return heap[0]


# Function to generate Huffman codes
def generate_huffman_codes(root, code, huffman_codes):
    if root is None:
        return
    if root.char is not None:
        huffman_codes[root.char] = code
    generate_huffman_codes(root.left, code + "0", huffman_codes)
    generate_huffman_codes(root.right, code + "1", huffman_codes)


# Function to perform Huffman coding
def huffman_coding(text):
    freq_map = defaultdict(int)
    for char in text:
        freq_map[char] += 1

    root = build_huffman_tree(freq_map)
    huffman_codes = {}
    generate_huffman_codes(root, "", huffman_codes)

    encoded_text = "".join(huffman_codes[char] for char in text)
    return encoded_text, root


# Example usage:
if __name__ == "__main__":
    text = "this is an example for huffman encoding"
    encoded_text, root = huffman_coding(text)
    print("Huffman Codes:")
    for char, code in sorted(huffman_codes.items()):
        print(f"{char}: {code}")
    print("Encoded Text:", encoded_text)
